src/app/registration/registration.component.spec.ts

```diff
@@ -44,4 +50,202 @@ describe('RegistrationComponent', () => {
   it('should create', () => {
     expect(component).toBeTruthy();
   });
+
+  // the following fullname field tests are to verify validaters required minlength pattern work
+
+  it('should validate required fullname field value is present ', () => {
+    let errors = {};
+    let fullnameFormControl = component.registerForm.get('fullname');
+    fullnameFormControl.setValue('Firstname Lastname');
+    errors = fullnameFormControl.errors || {};
+    expect(errors['required']).toBeUndefined(); // no errors
+  });
+
+  it('should validate required fullname field value is missing ', () => {
+    let errors = {};
+    let fullnameFormControl = component.registerForm.get('fullname');
+    fullnameFormControl.reset();
+    errors = fullnameFormControl.errors || {};
+    expect(errors['required']).toBeTruthy(); // has errors
+  });
+
+  it('should validate fullname field value contains at least 5 characters', () => {
+    let errors = {};
+    let fullnameFormControl = component.registerForm.get('fullname');
+    fullnameFormControl.setValue("test");
+    errors = fullnameFormControl.errors || {};
+    expect(errors['minlength']).toBeTruthy(); // has errors
+  });
+
+  it('should validate fullname field value may contains 5 or more characters', () => {
+    let errors = {};
+    let fullnameFormControl = component.registerForm.get('fullname');
+    fullnameFormControl.setValue("testing");
+    errors = fullnameFormControl.errors || {};
+    expect(errors['minlength']).toBeUndefined(); //no errors
+  });
+
+  it('should validate fullname field value is well formed (i.e., firstname space lastname)', () => {
+    let errors = {};
+    let fullnameFormControl = component.registerForm.get('fullname');
+    fullnameFormControl.setValue("Firstname Lastname");
+    errors = fullnameFormControl.errors || {};
+    expect(errors['pattern']).toBeUndefined(); // no error
+  });
+
+  it('should validate fullname field value is badly formed (i.e., firstname noSpace lastname)', () => {
+    let errors = {};
+    let fullnameFormControl = component.registerForm.get('fullname');
+    fullnameFormControl.setValue("FirstnameNoSpaceLastname");
+    errors = fullnameFormControl.errors || {};
+    expect(errors['pattern']).toBeTruthy(); // has errors
+  });
+
+  // the following email field tests are to verify validaters email required and asyncValidator emailTaken work
+
+  it('should validate required email field value is present ', () => {
+    let errors = {};
```

Comment:
I think there is a bit of a cut and paste error in this test.. I think it is not validating the value of the email field as you intended..

Reply:
I didn't check all the tests, but you should, to see if there are any other similar errors, or tests that are not doing what you intended..

Reply:
Johnathan (@haxwell) 

Wow, MyBad!!! I apologize for not catching the copy/paste errors previously. Shame on me for wasting your time. I think all those errors have been fixed now. Or at least I hope so ...

I also merged the recent dev change into this branch/PR.

Deborah

Reply:
No problem! I did not consider it a waste of time..